### ISSUE TITLE
Timeout with complex datatype

### ISSUE THREAD

**Issue by jafuehre-froeling** [dbus-cxx/dbus-cxx#113]
Hello!

First of all: thank you for your great work with this library!

I am making lots of successful DBUS-calls with the lib, but now I am pretty sure I did find some bug/limitation with your library.
The DBUS method that fails is `org.freedesktop.resolve1.Manager.RegisterService` (https://www.freedesktop.org/software/systemd/man/org.freedesktop.resolve1.html). When calling this method I get a `ErrorNoReply` (`Did not receive a response in the alotted time`).

Here is my code broken down to the relevant parts:
```cpp
// In my ::DBus::ObjectProxy
std::shared_ptr<::DBus::MethodProxy<::DBus::Path(std::string, std::string, std::string, uint16_t, uint16_t, uint16_t, std::vector<std::vector<std::map<std::string, std::vector<uint8_t>>>>)>> m_methodRegisterService;
void Setup() {
	m_methodRegisterService = this->create_method<::DBus::Path(std::string, std::string, std::string, uint16_t, uint16_t, uint16_t, std::vector<std::vector<std::map<std::string, std::vector<uint8_t>>>>)>("org.freedesktop.resolve1.Manager", "RegisterService");
}

inline static std::vector<uint8_t> StrToBinary(const std::string& data)
{
	return std::vector<uint8_t>(data.begin(), data.end());
}
int RegisterService() {
    std::vector<std::vector<std::map<std::string, std::vector<uint8_t>>>> txt_data =
	    {{{
		    {"key1", StrToBinary("value1")},
		    {"key2", StrToBinary("value2")},
		    {"key3", StrToBinary("value3")}
	    }}};
    
    ::DBus::Path result;
    
    try
    {
	    result = (*m_methodRegisterService)
		    ("test" // name
		    ,"%m" // name_template
		    , "_http._tcp" // type
		    , 0 // service_port
		    , 0 // service priority
		    , 0 //service_weight
		    , txt_datas);
    }
    catch (const ::DBus::Error& ex)
    {
	    LOGGER_CRITICAL("Caught exception: {0}: {1}", ex.name(), ex.message());
	    return -1;
    }
}
```

And here is the log output:
```
DBus::MethodProxy<N4DBus4PathE(NSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE,NSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE,NSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE,t,t,t,St6vectorIS_ISt3mapINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEES_IhSaIhEESt4lessIS6_ESaISt4pairIKS6_S8_EEESaISF_EESaISH_EE)> calling method=RegisterService
Creating call message to org.freedesktop.resolve1 path: /org/freedesktop/resolve1 org.freedesktop.resolve1.Manager.RegisterService
Signature 'o' is valid
Signature 's' is valid
Signature 's' is valid
Signature 's' is valid
Signature 's' is valid
Signature 'g' is valid
Signature 's' is valid
Signature 'ss' is valid
Signature 'g' is valid
Signature 'ss' is valid
Signature 'sss' is valid
Signature 'g' is valid
Signature 'sss' is valid
Signature 'sssq' is valid
Signature 'g' is valid
Signature 'sssq' is valid
Signature 'sssqq' is valid
Signature 'g' is valid
Signature 'sssqq' is valid
Signature 'sssqqq' is valid
Signature 'g' is valid
Signature 'a{say}' is valid
Signature '{say}' is valid
Signature 'av' is valid
Signature 'sssqqq' is valid
Signature 'sssqqqaav' is valid
Signature 'g' is valid
Signature 'a{say}' is valid
Signature '{say}' is valid
Signature 'y' is valid
Signature 'y' is valid
Signature 'y' is valid
Dispatching connections
Going to send the following bytes:
000000: 42 01 00 01 00 00 00 82   00 00 00 0b 00 00 00 a7   B.......  ........
000010: 01 01 6f 00 00 00 00 19   2f 6f 72 67 2f 66 72 65   ..o.....  /org/fre
000020: 65 64 65 73 6b 74 6f 70   2f 72 65 73 6f 6c 76 65   edesktop  /resolve
000030: 31 00 00 00 00 00 00 00   02 01 73 00 00 00 00 20   1.......  ..s....
000040: 6f 72 67 2e 66 72 65 65   64 65 73 6b 74 6f 70 2e   org.free  desktop.
000050: 72 65 73 6f 6c 76 65 31   2e 4d 61 6e 61 67 65 72   resolve1  .Manager
000060: 00 00 00 00 00 00 00 00   03 01 73 00 00 00 00 0f   ........  ..s.....
000070: 52 65 67 69 73 74 65 72   53 65 72 76 69 63 65 00   Register  Service.
000080: 06 01 73 00 00 00 00 18   6f 72 67 2e 66 72 65 65   ..s.....  org.free
000090: 64 65 73 6b 74 6f 70 2e   72 65 73 6f 6c 76 65 31   desktop.  resolve1
0000a0: 00 00 00 00 00 00 00 00   08 01 67 00 09 73 73 73   ........  ..g..sss
0000b0: 71 71 71 61 61 76 00 00   00 00 00 04 74 65 73 74   qqqaav..  ....test
0000c0: 00 00 00 00 00 00 00 02   25 6d 00 00 00 00 00 0a   ........  %m......
0000d0: 5f 68 74 74 70 2e 5f 74   63 70 00 00 00 00 00 00   _http._t  cp......
0000e0: 00 00 00 00 00 00 00 52   00 00 00 4e 00 00 00 46   .......R  ...N...F
0000f0: 00 00 00 00 00 00 00 04   6b 65 79 31 00 00 00 00   ........  key1....
000100: 00 00 00 06 76 61 6c 75   65 31 00 00 00 00 00 04   ....valu  e1......
000110: 6b 65 79 32 00 00 00 00   00 00 00 06 76 61 6c 75   key2....  ....valu
000120: 65 32 00 00 00 00 00 04   6b 65 79 33 00 00 00 00   e2......  key3....
000130: 00 00 00 06 76 61 6c 75   65 33                     ....valu  e3
Try to read a message
done dispatching
```

A manual call via `gdbus call --system --dest org.freedesktop.resolve1 --object-path /org/freedesktop/resolve1 -m org.freedesktop.resolve1.Manager.RegisterService "test" "%m" "_http._tcp" 0 0 0
 "[{'key1': [118,97,108,117,101,49]}]"` works.
 
 I checked this out with gdb, the return value of https://github.com/dbus-cxx/dbus-cxx/blob/dfa9dc56fc70dec9c4f6b7dc9a885cd046605244/dbus-cxx/sendmsgtransport.cpp#L328 matches size of the message to be sent.

Please contact me if you need more information on this!

Ps: Increasing the timeout does not fix this.
PPs: Version 2.3.1 is used

**Comment by rm5248** [comment]
The most likely reason this is failing is because the parameters are incorrect.

The signature for the `txt_datas` parameters is `aa{say}`.  you have:
```
std::vector<std::vector<std::map<std::string, std::vector<uint8_t>>>>
```

It looks like you have an extra `std::vector` in there - maps are shown as `array{key,value}` in introspection code, so this should probably be:
```
std::vector<std::map<std::string,std::vector<uint8_t>>>
```

**Comment by jafuehre-froeling** [comment]
You were right, chaning the signature fixed the timeout issue! I did misunderstand the dbus dict type. Unfortunately, now I get a org.freedesktop.DBus.Error.NotSupported and I really don't understand why. I watched with `dbus-monitor` and compared the message with the `gdbus`-generated one and they are equal. I cannot explain this difference. I looked through the source of systemd and dbus and this error has to come from the OS. Running on a armv7l.

```
Try to read a message
done dispatching
Dispatching connections
Try to read a message
Have 24 bytes of control after real reading
Signature 'u' is valid
Signature 'u' is valid
Signature 's' is valid
Signature 's' is valid
Signature 's' is valid
Signature 's' is valid
Signature 'g' is valid
Signature 's' is valid
Signature 'g' is valid
Signature 's' is valid
Signature 's' is valid
Creating ErrorMessage from data
Message has 0 fds
Signature 's' is valid
Following message created from the data: DBus::Message = [ErrorMessage
  Message length: 28
  Endianess: Little
  Serial: 29
  Headers:
    Error Name: org.freedesktop.DBus.Error.NotSupported
    Reply Serial: 11
    Destination: :1.22
    Sender: :1.4
    Signature: DBus::Signature[s]
]
done dispatching
Signature 's' is valid
Signature 's' is valid
```

Do you have an idea where this difference can come from?

**Comment by rm5248** [comment]
If you can post the output of dbus-monitor using both dbus-cxx and gdbus that would be the best way to compare them.  Probably something is still misspelled.

**Comment by jafuehre-froeling** [comment]
I compared the two dbus-monitor results, they are exactly the same, except for the response. I really cannot comprehend where this difference comes from. I read though systemd source code https://github.com/systemd/systemd/blob/9eba03c7b10fe98e0443508402b3f9804230453e/src/resolve/resolved-bus.c#L1881, the only thing that might cause the difference might be the euid check?

**Comment by rm5248** [comment]
Without having a dump of the exact messages that you are seeing I don't know what it could be.

If you can capture some traffic with dbus-monitor as a PCAP file I can look at it(use `dbus-monitor --pcap`)

**Comment by jafuehre-froeling** [comment]
I've created a minimal test executable and that yields the same result. Here are two PCAP-files, one capturing gdbus-calls and one dbus-cxx-calls.
[captures.tar.gz](https://github.com/dbus-cxx/dbus-cxx/files/11259264/captures.tar.gz)

**Comment by jafuehre-froeling** [comment]
Here is the minimal source code:

```cpp
#include <iostream>
#include <dbus-cxx.h>

constexpr auto RESOLVED_DEST = "org.freedesktop.resolve1";
constexpr auto RESOLVED_NODE = "/org/freedesktop/resolve1";
constexpr auto RESOLVED_MANAGER_INTERFACE = "org.freedesktop.resolve1.Manager";

class ResolvedManagerObject : public ::DBus::ObjectProxy
{
protected:
	explicit ResolvedManagerObject(std::shared_ptr<::DBus::Connection> connection)
			: ::DBus::ObjectProxy(std::move(connection), RESOLVED_DEST, RESOLVED_NODE)
	{
		m_methodRegisterService = this->create_method<::DBus::Path(std::string, std::string, std::string, uint16_t, uint16_t, uint16_t, std::vector<std::map<std::string, std::vector<uint8_t>>>)>(RESOLVED_MANAGER_INTERFACE, "RegisterService");
	}

public:
	static std::shared_ptr<ResolvedManagerObject> Create(const std::shared_ptr<::DBus::Connection>& connection)
	{
		auto managerObject = std::shared_ptr<ResolvedManagerObject>(new ResolvedManagerObject(connection));
		connection->register_object_proxy(managerObject);
		return managerObject;
	}

	::DBus::Path RegisterService(const std::string& name, const std::string& name_template, const std::string& type, uint16_t service_port, uint16_t service_priority, uint16_t service_weight, const std::vector<std::map<std::string, std::vector<uint8_t>>>& txt_data)
	{
		return (*m_methodRegisterService)(name, name_template, type, service_port, service_priority, service_weight, txt_data);
	}

private:
	std::shared_ptr<::DBus::MethodProxy<::DBus::Path(std::string, std::string, std::string, uint16_t, uint16_t, uint16_t, std::vector<std::map<std::string, std::vector<uint8_t>>>)>> m_methodRegisterService;
};

inline static std::vector<uint8_t> StrToBinary(const std::string& data)
{
	return std::vector<uint8_t>(data.begin(), data.end());
}

int main(int argc, char* argv[])
{
	std::vector<std::map<std::string, std::vector<uint8_t>>> txt_data =
		{{{
			{"test1", StrToBinary("value1")},
			{"test2", StrToBinary("value2")},
			{"test3", StrToBinary("value3")}
		}}};

	::DBus::Path result;

	try
	{
		auto dispatcher = ::DBus::StandaloneDispatcher::create();
		auto connection = dispatcher->create_connection(::DBus::BusType::SYSTEM);
		auto managerObject = ResolvedManagerObject::Create(connection);
		result = managerObject->RegisterService("test", "%m", "_http._tcp", 0, 0, 0, txt_data);
	}
	catch (const ::DBus::Error& ex)
	{
		std::cerr << "main: Caught exception: " << ex.name() << ": " << ex.message() << std::endl;
		return -1;
	}
	catch (...)
	{
		std::exception_ptr p = std::current_exception();
		std::string ex = (p ? p.__cxa_exception_type()->name() : "null");
		std::cerr << "main: Caught exception: " << ex << std::endl;
		return -1;
	}

	std::cout << "Success: " << result << std::endl;
	return 0;
}
```

**Comment by rm5248** [comment]
Thanks!  A quick look at it indicates that there is something wrong with the message body.  From the CLI, the message body length is 78 bytes, while with dbus-cxx the total message body length is 126 bytes.  Possibly the error is actually coming from libdbus once it is trying to parse the message, but that seems odd as any errors should have been caught by dbus-daemon before forwarding on the message.

I should be able to take a look at it this weekend.

**Comment by jafuehre-froeling** [comment]
> the message body length is 78 bytes, while with dbus-cxx the total message body length is 126 bytes

Thats a red herring, the txtdata-data is slightly different. I'll rerun this without any txtdata to make this easier to compare.

**Comment by jafuehre-froeling** [comment]
[captures.tar.gz](https://github.com/dbus-cxx/dbus-cxx/files/11280194/captures.tar.gz)

Update: The call works when not providing txtdata. This gz contains two calls with `"key1": "value1"`. One key difference I noticed is the endianness

**Comment by rm5248** [comment]
I can confirm that this is because the endianess is different.  I did a quick test on my side changing all of the marshaling endianess to little, and I got a new error(`org.freedesktop.DBus.Error.InteractiveAuthorizationRequired`).

So it looks like the root cause is a bug in sdbus.

As a workaround, I will add the ability to change the endianess - I thought it would be more difficult than it is, but fortunately it seems that there is no demarshaling done in any header files, so this should be possible to fix without breaking anything or forcing a recompile.
